midp/gfx.js

```diff
@@ -98,10 +169,20 @@ var currentlyFocusedTextEditor;
 
     var refreshStr = "refresh";
     Native["com/sun/midp/lcdui/DisplayDevice.refresh0.(IIIIII)V"] = function(hardwareId, displayId, x1, y1, x2, y2) {
-        x1 = Math.max(0, x1);
-        y1 = Math.max(0, y1);
-        x2 = Math.max(0, x2);
-        y2 = Math.max(0, y2);
+        if (!config.useOffscreenCanvas) {
+            // If the offscreen canvas is disabled, then we draw directly
+            // to the device canvas, so it's always up-to-date and never needs
+            // to be refreshed.
+            return;
+        }
+
+        x1 = Math.max(0, x1) * DPR;
+        y1 = Math.max(0, y1) * DPR;
+        x2 = Math.max(0, x2) * DPR;
+        y2 = Math.max(0, y2) * DPR;
+
+        // XXX Isn't the offscreen and device canvas width and height always
+        // the same, so it isn't necessary to compute which one is more minimal?
```

Comment:
Erm, my comment is a bit misleading. I meant to say that the offscreen and device canvas widths should always be the same, and their heights should similarly be the same. So these calls should be unnecessary:

```JavaScript
        var maxX = Math.min(offscreenCanvas.width, MIDP.deviceContext.canvas.width);
        …
        var maxY = Math.min(offscreenCanvas.height, MIDP.deviceContext.canvas.height);
```

I don't think there's ever a situation where they can be different, based on what I know of the code. As soon as the device canvas size changes, a canvasresize event is dispatched, and the canvasresize handler resizes the offscreen canvas to match the device canvas's size.

But perhaps I'm missing something. And changing it would be orthogonal to this patch. So I've left the code as-is but clarified the comment to read:

```JavaScript
        // XXX Aren't the offscreen and device canvas always the same size,
        // so it isn't necessary to compute their minimum width and height?
```